classes/plugins/PubObjectsExportPlugin.inc.php

```diff
@@ -51,12 +51,25 @@ function getCache() {
 	 * @copydoc Plugin::register()
 	 */
 	function register($category, $path, $mainContextId = null) {
-		$success = parent::register($category, $path, $mainContextId);
-		if ($success) {
-			$this->addLocaleData();
-			HookRegistry::register('AcronPlugin::parseCronTab', array($this, 'callbackParseCronTab'));
+		if (!parent::register($category, $path, $mainContextId)) return false;
+
+		$this->addLocaleData();
+		HookRegistry::register('AcronPlugin::parseCronTab', array($this, 'callbackParseCronTab'));
+		foreach ($this->_getDAOs() as $dao) {
+			if ($dao instanceof SchemaDAO) {
+				$schema = Services::get('schema')->get($dao->schemaName);
+				foreach ($this->_getObjectAdditionalSettings() as $fieldName) {
+					$schema->properties->{$fieldName} = (object) [
+						'type' => 'string',
+						'apiSummary' => true,
+						'validation' => ['nullable'],
+					];
+				}
```

Comment:
Oh, this is clever. I hadn't thought about adding onto the schema object after it has been fetched from the json. But of course since its an object its a reference and that change will propagate throughout the system.

One possible reason to maybe avoid this as a general practice is that I think it will not integrate nicely with the REST API documentation. One of the benefits of the existing json/schema system is that if a plugin modifies a schema that will automatically be reflected in the swagger API specification that is generated, as long as the plugin is enabled when the swagger is generated.

It might be worth running a quick test against the API to see if these new properties are getting added to the API response for a `Publication` object when you have the plugin enabled. I'd be curious to know if they get added on from here. If so, that's even better, and the documentation thing is a very minor downside.

Reply:
Thanks, Nate -- 

Because the import/export plugins aren't loaded when the API is used, and likewise when the Swagger docs are generated, the added elements won't be seen in either place. I'd like to think a little further about this and see if we can't come up with a unified approach to extending the schema, but I suggest we defer that to e.g. 3.3 and treat this use case -- a non-generic plugin adding settings to things -- as a bit of an outlier. I suspect a clean solution that doesn't force all plugins to load all the time will probably involve some use of persistence that we don't yet have (e.g. the settings list maintained in the database, which risks being a mess), or some cleanup around plugin loading that removes the lazy-load tools, loads all plugins on all requests (!), but limits their actions unless enabled to an acceptable minimum.

Extending the schema this way feels cleaner to me than using the hook, but I'd rather not introduce that kind of change this late in the game.

Are you OK with this approach for the 3.2 release?

Reply:
(I already got a yes but missed it! https://github.com/pkp/pkp-lib/issues/5474#issuecomment-591331085)